django/db/backends/sqlite3/_functions.py

```diff
@@ -0,0 +1,471 @@
+    create_deterministic_function('SHA224', 1, _sqlite_sha224)
+    create_deterministic_function('SHA256', 1, _sqlite_sha256)
+    create_deterministic_function('SHA384', 1, _sqlite_sha384)
+    create_deterministic_function('SHA512', 1, _sqlite_sha512)
+    create_deterministic_function('SIGN', 1, _sqlite_sign)
+    create_deterministic_function('SIN', 1, _sqlite_sin)
+    create_deterministic_function('SQRT', 1, _sqlite_sqrt)
+    create_deterministic_function('TAN', 1, _sqlite_tan)
+    # Don't use the built-in RANDOM() function because it returns a value
+    # in the range [-1 * 2^63, 2^63 - 1] instead of [0, 1).
+    connection.create_function('RAND', 0, random.random)
+    connection.create_aggregate('STDDEV_POP', 1, StddevPop)
+    connection.create_aggregate('STDDEV_SAMP', 1, StddevSamp)
+    connection.create_aggregate('VAR_POP', 1, VarPop)
+    connection.create_aggregate('VAR_SAMP', 1, VarSamp)
+
+
+def _sqlite_datetime_parse(dt, tzname=None, conn_tzname=None):
+    if dt is None:
+        return None
+    try:
+        dt = typecast_timestamp(dt)
+    except (TypeError, ValueError):
+        return None
+    if conn_tzname:
+        dt = dt.replace(tzinfo=timezone_constructor(conn_tzname))
+    if tzname is not None and tzname != conn_tzname:
+        tzname, sign, offset = split_tzname_delta(tzname)
+        if offset:
+            hours, minutes = offset.split(':')
+            offset_delta = timedelta(hours=int(hours), minutes=int(minutes))
+            dt += offset_delta if sign == '+' else -offset_delta
+        dt = timezone.localtime(dt, timezone_constructor(tzname))
+    return dt
+
+
+def _sqlite_date_trunc(lookup_type, dt, tzname, conn_tzname):
+    dt = _sqlite_datetime_parse(dt, tzname, conn_tzname)
+    if dt is None:
+        return None
+    if lookup_type == 'year':
+        return "%i-01-01" % dt.year
+    elif lookup_type == 'quarter':
+        month_in_quarter = dt.month - (dt.month - 1) % 3
+        return '%i-%02i-01' % (dt.year, month_in_quarter)
+    elif lookup_type == 'month':
+        return "%i-%02i-01" % (dt.year, dt.month)
+    elif lookup_type == 'week':
+        dt = dt - timedelta(days=dt.weekday())
+        return "%i-%02i-%02i" % (dt.year, dt.month, dt.day)
+    elif lookup_type == 'day':
+        return "%i-%02i-%02i" % (dt.year, dt.month, dt.day)
+
+
+def _sqlite_time_trunc(lookup_type, dt, tzname, conn_tzname):
+    if dt is None:
+        return None
+    dt_parsed = _sqlite_datetime_parse(dt, tzname, conn_tzname)
+    if dt_parsed is None:
+        try:
+            dt = typecast_time(dt)
+        except (ValueError, TypeError):
+            return None
+    else:
+        dt = dt_parsed
+    if lookup_type == 'hour':
+        return "%02i:00:00" % dt.hour
+    elif lookup_type == 'minute':
+        return "%02i:%02i:00" % (dt.hour, dt.minute)
+    elif lookup_type == 'second':
+        return "%02i:%02i:%02i" % (dt.hour, dt.minute, dt.second)
```

Comment:
Various things that can be improved here:

- Use f-strings for a speed boost over %-formatting - could add up significantly
- Avoid unnecessary `elif`s.
- Raise `NotImplementedError` in the default case for unknown lookups.

```suggestion
    if lookup_type == 'hour':
        return f'{dt.hour:02d}:00:00'
    if lookup_type == 'minute':
        return f'{dt.hour:02d}:{dt.minute:02d}:00'
    if lookup_type == 'second':
        return f'{dt.hour:02d}:{dt.minute:02d}:{dt.second:02d}'
    raise NotImplementedError(f'Unsupported lookup type: {lookup_type}')
```

Alternatively it is possible to do the following, but it is likely to be significantly slower as I think it delegates to `.strftime()` under the hood:

```suggestion
    if lookup_type == 'hour':
        return f'{dt:%H}:00:00'
    if lookup_type == 'minute':
        return f'{dt:%H:%M}:00'
    if lookup_type == 'second':
        return f'{dt:%H:%M:%S}'
    raise NotImplementedError(f'Unsupported lookup type: {lookup_type}')
```